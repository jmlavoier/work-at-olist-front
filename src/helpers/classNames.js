const classNames = stylesConditions => Object.keys(stylesConditions).reduce(
  (className, item) => {
    if (stylesConditions[item]) {
      return `${className} ${item}`
    }

    return className;
  }
, '');

export default classNames;
