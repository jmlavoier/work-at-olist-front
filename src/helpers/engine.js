export class Component {
  constructor(props = {}) {
    this.childrenComponents = [];
    this.props = props;
    this.id = `${this.getRandomId()}`;
    this.state = {};
    this.isMounted = true;

    this.events = [];
    this.listeners = [];

    this.componentWillMount();
    this.render();
    this.componentDidMount();
    this.subscribeEvents();
    this.executeListeners({}, this.state);
  }

  componentWillMount() {}
  componentDidMount() {}
  componentWillUnmount() {}
  render() {}

  setState(state) {
    let nextState = {};
    const previousState = { ...this.state };

    if (typeof state === 'function') {
      nextState = { ...state(this.state) };
    } else {
      nextState = { ...state };
    }

    this.state = {
      ...previousState,
      ...nextState,
    }

    this.executeListeners(previousState, this.state);
  }

  executeListeners(previousState, nextState) {
    this.listeners.forEach(listener => listener(previousState, nextState));
  }

  eventHandler(componentEvent) {
    return (event) => {
      componentEvent(event);
    }
  }

  subscribeEvents() {
    this.events.forEach((componentEvents) => {
      const { el, events } = componentEvents;
      const eventsName = Object.keys(events);

      if (!el) {
        return;
      }

      eventsName.forEach((event) => {
        el.addEventListener(event, this.eventHandler(events[event]), true);
      });
    });
  }

  getRandomId() {
    return `${parseInt(123456 * Math.random(), 10)}`;
  }

  unmountComponent() {
    if (this.isMounted) {
      this.componentWillUnmount();
      this.isMounted = false;

      if (this.parentComponent) {
        this.parentComponent.renderChildrenUnmontedComponents();
      }
    }
  }

  mountComponent() {
    this.componentWillMount();
    this.render();
    this.componentDidMount();
    this.subscribeEvents();
    this.executeListeners({}, this.state);
    this.isMounted = true;

    if (this.parentComponent) {
      this.parentComponent.renderChildrenComponents();
    }
  }

  getChildComponentRefByName(name) {
    const [compRef] = this.childrenComponents.filter(
      (componentRef) => componentRef.name === name
    );

    if (compRef) {
      return compRef;
    }

    return null;
  }

  getTmpSpan() {
    const tmpSpan = `<span id="tmp-${this.id}"></span>`;
    return this.parseElement(tmpSpan);
  }

  template(strings, ...children) {
    let str = '';
    this.childrenComponents = [];

    const templateString = strings.reduce((html, item, index) => {
      const child = children[index];

      if (child && child.component && typeof child.component === 'function') {
        const childComponent = new child.component(child.props);
        childComponent.parentComponent = this;

        if (childComponent instanceof Component) {
          const componentRef = this.getChildComponentRefByName(child.name) || {
            ...child,
            startedComponent: childComponent,
          };

          this.childrenComponents.push(componentRef);
          return `${html}${item} <span id="tmp-${componentRef.startedComponent.id}"></span>`;
        }
      }

      if (typeof child === 'boolean') {
        return `${html} ${item}`;
      }

      if (typeof child === 'array') {
        throw new Error('Component doesn\'t support array as child');
      }

      if (typeof child === 'number') {
        return `${html} ${item} ${child.toString() || ''}`;
      }

      return `${html}${item}${child || ''}`;
    }, '');

    this.element = this.parseElement(templateString);
    this.element.id = `cmp-${this.id}`
  }

  parseElement(templateString) {
    const html = document.createElement('div');
    html.innerHTML = templateString;
    return html.firstElementChild;
  }

  getElement() {
    return this.element || '';
  }

  renderChildrenComponents() {
    this.childrenComponents.forEach((componentRef) => {
      if (componentRef.startedComponent.isMounted) {
        const tmpSpan = this.element.querySelector(`#tmp-${componentRef.startedComponent.id}`);

        if (tmpSpan) {
          tmpSpan.replaceWith(componentRef.startedComponent.getElement());
          componentRef.startedComponent.renderChildrenComponents();
        }
      }
    });
  }

  renderChildrenUnmontedComponents() {
    this.childrenComponents.forEach((componentRef) => {
      if (!componentRef.startedComponent.isMounted) {
        const compDOM = this.element.querySelector(`#cmp-${componentRef.startedComponent.id}`);

        if (compDOM) {
          compDOM.replaceWith(componentRef.startedComponent.getTmpSpan());
          componentRef.startedComponent.renderChildrenComponents();
        }
      }
    });
  }

  renderRoot(root) {
    root.appendChild(this.element);
    if (this.childrenComponents.length) {
      this.renderChildrenComponents();
      this.renderChildrenUnmontedComponents();
    }
  }
}

export const prepareComponent = component => (name, props) => {
  if (!name) {
    throw new Error('Component should have name to be prepared');
  }

  return {
    name,
    props,
    component,
  };
}

//For tests
export const mount = (Component) => {
  const html = document.createElement('div');
  Component.element.id = '';
  const spans = Component.element.querySelectorAll('span[id^="tmp-"]');
  if (spans) {
    for (let i = 0; i < spans.length; i++) {
      spans[i].id = '';
    }
  }
  html.appendChild(Component.element);
  return {
    el: html.firstElementChild,
    string: html.innerHTML,
  };
}