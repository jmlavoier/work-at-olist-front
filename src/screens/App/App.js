import React from 'react';

import Success from 'screens/Success';
import styles from './App.sass';

const App = () => (
  <div className={styles.container}>
    <div className={styles.box}>
      <Success />
    </div>
  </div>
);

export default App;
