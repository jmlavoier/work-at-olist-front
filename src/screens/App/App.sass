.container
  display: flex
  margin-top: 80px
  justify-content: center

.box
  width: 100%
  max-width: 515px