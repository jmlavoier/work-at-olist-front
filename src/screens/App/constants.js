export const SCREENS = {
  form: 'Form',
  success: 'Success',
};

export default SCREENS;
