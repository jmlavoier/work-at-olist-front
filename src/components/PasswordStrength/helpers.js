import classNames from 'classnames';
import styles from './PasswordStrength.sass';

export const createRules = value => ({
  hasMinChar: value.length >= 6,
  hasUppercase: /[A-Z]+/.test(value),
  hasNumber: /[0-9]+/.test(value),
});

export const getStrengthvalue = (rules) => {
  const {
    hasMinChar,
    hasUppercase,
    hasNumber,
  } = rules;

  return (hasMinChar ? 1 : 0) + (hasUppercase ? 1 : 0) + (hasNumber ? 1 : 0);
};

export const getStylesBar1 = strengthValue => classNames({
  [styles.gray]: strengthValue === 0,
  [styles.red]: strengthValue === 1,
  [styles.yellow]: strengthValue === 2,
  [styles.green]: strengthValue === 3,
});

export const getStylesBar2 = strengthValue => classNames({
  [styles.gray]: strengthValue <= 1,
  [styles.yellow]: strengthValue === 2,
  [styles.green]: strengthValue === 3,
});

export const getStylesBar3 = strengthValue => classNames({
  [styles.gray]: strengthValue <= 2,
  [styles.green]: strengthValue === 3,
});

export const getStylesRule = (value, attendRule) => classNames(styles.rule, {
  [styles['rule-red']]: value && !attendRule,
  [styles['rule-green']]: value && attendRule,
});

export const styleSheet = {
  inputText: {
    width: '100%',
    marginBottom: 0,
  },
};
