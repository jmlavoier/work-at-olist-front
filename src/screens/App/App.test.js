import { prepareComponent } from 'helpers/engine';
import App from './App';

const getDOM = (Component) => {
  Component.renderChildrenComponents();
  const html = document.createElement('div');
  html.appendChild(Component.element);
  return {
    el: html.firstElementChild,
    string: html.innerHTML,
  };
}

describe('<App />', () => {
  it('Should component match snapshot', () => {
    const wrapper = getDOM(App);

    expect(wrapper.string).toMatchSnapshot();
  });
});
