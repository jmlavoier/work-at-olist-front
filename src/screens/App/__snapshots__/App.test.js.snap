// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Should component match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="box"
  >
    <Success />
  </div>
</div>
`;
